torch/distributed/fsdp/sharded_grad_scaler.py

```diff
@@ -281,27 +281,35 @@ def unscale_(self, optimizer: torch.optim.Optimizer) -> None:
         # Synchronize the detected inf across the ranks
         optimizer_state = self._per_optimizer_states[id(optimizer)]
         future_handles = []
-
-        for v in optimizer_state["found_inf_per_device"].values():
-            if v.device.type == "cpu":
-                v_on_cuda = v.cuda()
+        found_inf_on_cpus = []
+        found_inf_on_cudas = []
+
+        for found_inf in optimizer_state["found_inf_per_device"].values():
+            if found_inf.device.type == "cpu":
+                found_inf_on_cpus.append(found_inf)
+                found_inf_on_cuda = found_inf.cuda()
+                found_inf_on_cudas.append(found_inf_on_cuda)
                 future_handles.append(
                     dist.all_reduce(
-                        v_on_cuda, async_op=True, group=self.process_group
+                        found_inf_on_cuda, async_op=True, group=self.process_group
                     ).get_future()
                 )
-                v.copy_(v_on_cuda.cpu())
             else:
                 future_handles.append(
                     dist.all_reduce(
-                        v, async_op=True, group=self.process_group
+                        found_inf, async_op=True, group=self.process_group
                     ).get_future()
                 )
 
         # Make sure that the calls are done before moving out.
         if future_handles:
             torch.futures.wait_all(future_handles)
 
+        for found_inf_on_cpu, found_inf_on_cuda in zip(
+            found_inf_on_cpus, found_inf_on_cudas
+        ):
+            found_inf_on_cpu.copy_(found_inf_on_cuda)
```

Comment:
Can consider
```suggestion
    torch._foreach_copy_(found_inf_on_cpus, found_inf_on_cudas)
```
to save Python-C++ context switching